platform-sdk/swirlds-merkledb/src/main/java/com/swirlds/merkledb/files/hashmap/HalfDiskHashMap.java

```diff
@@ -425,93 +465,242 @@ public DataFileReader endWriting() throws IOException {
+                throw new RuntimeException("Exception in HDHM.endWriting()", z);
             }
-            // close files session
-            dataFileReader = fileCollection.endWriting(0, numOfBuckets);
-            // we have updated all indexes so the data file can now be included in merges
-            dataFileReader.setFileCompleted();
         } else {
             dataFileReader = null;
         }
-
         // clear put cache
         oneTransactionsData = null;
         return dataFileReader;
     }
 
     /**
-     * Reads a bucket with a given index from disk, updates given keys in it, and puts the bucket to
-     * a queue. If an exception is thrown, it's put to the queue instead, so the number of {@code
-     * ReadBucketResult} objects in the queue is consistent.
-     *
-     * @param bucketIndex The bucket index
-     * @param keyUpdates Key/value updates to apply to the bucket
-     * @param queue The queue to put the bucket or exception to
+     * A task to submit "read bucket" tasks. Tasks are submitted till the number of buckets
+     * in progress exceeds MAX_IN_FLIGHT. After that, if there are still unprocessed buckets,
+     * a new "submit task" is scheduled. This new task is run after at least one bucket is
+     * fully processed, i.e. its "store" task is complete.
      */
-    private void readUpdateQueueBucket(
-            final int bucketIndex, final BucketMutation keyUpdates, final Queue<ReadBucketResult> queue) {
-        try {
-            // The bucket will be closed on the lifecycle thread
-            final Bucket bucket;
-            BufferedData bucketData = fileCollection.readDataItemUsingIndex(bucketIndexToBucketLocation, bucketIndex);
-            if (bucketData == null) {
-                // create a new bucket
-                bucket = bucketPool.getBucket();
-                bucket.setBucketIndex(bucketIndex);
+    private class SubmitTask extends AbstractTask {
+
+        private final Iterator<IntObjectPair<BucketMutation>> it;
+
+        SubmitTask(final ForkJoinPool pool, final Iterator<IntObjectPair<BucketMutation>> it, final int depCount) {
+            super(pool, depCount);
+            this.it = it;
+        }
+
+        // Notifies that some bucket is fully processed. It sets one of this task's dependencies.
+        // The other one is set in the end of the previous submit task
+        void notifyBucketProcessed() {
+            send();
+        }
+
+        @Override
+        protected boolean exec() {
+            // The next submit task to run after the current one. It will only be run, if
+            // this task doesn't schedule tasks for all remaining buckets, and at least one
+            // bucket is completely processed while this method is running
+            final SubmitTask nextSubmitTask = new SubmitTask(getPool(), it, 2);
+            final boolean newSubmitTaskSet = currentSubmitTask.compareAndSet(this, nextSubmitTask);
+            assert newSubmitTaskSet;
```

Comment:
`currentSubmitTask` does not need to be atomic, not even `volatile`. JMM ensures its visibility in all threads due to atomic updates of `bucketPermits` (that probably needs to be explicitly stated in comments).

Reply:
Discussed it with Oleg offline, this is not critical (and the current version is more readable)